rat/variable_table.py

```diff
@@ -0,0 +1,242 @@
+        return self.symbol_dict[variable_name]
+
+    def __iter__(self):
+        for name in self.symbol_dict:
+            yield name
+
+    def prepare_unconstrained_parameter_mapping(self):
+        """
+        Compute the size of vector that can hold everything and figure out
+        how each record maps into this vector
+        """
+        current_index = 0
+
+        for variable_name, record in self.symbol_dict.items():
+            if not record.variable_type:
+                error_msg = f"Fatal internal error - variable type information for '{variable_name}' not present within the symbol table. Aborting compilation."
+                raise CompileError(error_msg)
+
+            # Allocate space on unconstrained parameter vector for parameters
+            if record.variable_type == VariableType.PARAM:
+                if len(record.subscripts) > 0:
+                    nrows = record.base_df.shape[0]
+                else:
+                    nrows = 1
+
+                record.unconstrained_vector_start_index = current_index
+                record.unconstrained_vector_end_index = current_index + nrows - 1
+                current_index += nrows
+
+        self.unconstrained_parameter_size = current_index
+
+    def get_subscript_key(
+        self,
+        primary_variable_name: str,
+        target_variable_name: str,
+        target_variable_subscripts: Tuple[str],
+        shifts: Tuple[int],
+    ) -> str:
+        primary_variable = self[primary_variable_name]
+        target_variable = self[target_variable_name]
+
+        shifts_by_subscript_name = {name: shift for name, shift in zip(target_variable_subscripts, shifts)}
+
+        target_variable.base_df
+
+        shift_subscripts = []
+        shift_values = []
+        grouping_subscripts = []
+        for column in primary_variable.base_df.columns:
+            if column in shifts_by_subscript_name:
+                shift = shifts_by_subscript_name[column]
+            else:
+                shift = 0
+
+            if shift == 0:
+                grouping_subscripts.append(column)
+            else:
+                shift_subscripts.append(column)
+                shift_values.append(shift)
+
+        primary_base_df = primary_variable.base_df.copy()
+
+        if len(grouping_subscripts) > 0:
+            grouped_df = primary_base_df.groupby(grouping_subscripts)
+        else:
+            grouped_df = primary_base_df
+
+        for column, shift in zip(shift_subscripts, shift_values):
+            shifted_column = grouped_df[column].shift(shift).reset_index(drop=True)
+            primary_base_df[column] = shifted_column
+
+        target_base_df = target_variable.base_df.copy()
+        target_base_df.columns = list(target_variable_subscripts)
+
+        target_base_df["__in_dataframe_index"] = pandas.Series(range(target_base_df.shape[0]), dtype=pandas.Int64Dtype())
+
+        key_name = f"subscript__{self.get_unique_number()}"
```

Comment:
Looking at this made me remember something: do you think it would be worth "reusing" generated index arrays? Right now it generates a new index array every time `get_subscript_key` is called, but it can be changed so that if we already created indices that share the same primary and target variables, we return the name of the previously calculated array. 

Of course, that's added complexity to an already complicated function and I'm dubious if it's worth doing.

Reply:
Hmm I guess it's possible but yeah I favor simplicity here until performance is a problem.